storage-transcribe-audio/functions/src/transcribe-audio.ts

```diff
@@ -0,0 +1,212 @@
+    return {
+      state: "failure",
+      warnings,
+      type: FailureType.NULL_TRANSCRIPTION,
+    };
+  }
+
+  // I simplify the transcription compared to the one that's usually given
+  // by the cloud call because, for example, we don't give the option
+  // to request many candidate transcriptions from speech to text.
+  //
+  // However, the simplification doesn't happen for the file uploaded by the API
+  // Should I
+  // (a) just not simplify?
+  // (b) simplify after the upload has succeeded?
+  // (c) simplify, then upload (without harnessing the upload of the cloud API)?
+  // (d) simplify, keeping the mildly inconsistent behavior?
+  const transcription = separateByTags(taggedTranscription);
+
+  logs.logResponseTranscription(transcription);
+  return {
+    state: "success",
+    warnings,
+    transcription,
+  };
+}
+
+export async function transcodeToLinear16AndUpload(
+  {
+    localCopyPath,
+    storageOutputPath,
+  }: { localCopyPath: string; storageOutputPath: string },
+  bucket: Bucket
+): Promise<TranscodeAudioResult> {
+  const probeData: ffmpeg.FfprobeData = await probePromise(localCopyPath);
+  const warnings: WarningType[] = [];
+
+  logs.debug("probe data before transcription:", probeData);
+  const { streams } = probeData;
+
+  if (streams.length === 0) {
+    return {
+      state: "failure",
+      type: FailureType.ZERO_STREAMS,
+      warnings,
+    };
+  }
+
+  if (streams.length !== 1) {
+    warnings.push(WarningType.MORE_THAN_ONE_STREAM);
+  }
+
+  if (streams[0].sample_rate == null) {
+    return {
+      state: "failure",
+      type: FailureType.NULL_SAMPLE_RATE,
+      warnings,
+    };
+  }
+
+  if (streams[0].channels == null) {
+    return {
+      state: "failure",
+      type: FailureType.NULL_CHANNELS,
+      warnings,
+    };
+  }
+
+  const localOutputPath = localCopyPath + TRANSCODE_TARGET_FILE_EXTENSION;
+  logs.debug("transcoding locally");
+  try {
+    await transcodeLocally({
+      inputPath: localCopyPath,
+      outputPath: localOutputPath,
+    });
+  } catch (error: unknown) {
+    const { err, stdout, stderr } = error as {
+      err: any;
+      stdout: string;
+      stderr: string;
+    };
+    return {
+      state: "failure",
+      type: FailureType.FFMPEG_FAILURE,
+      warnings,
+      details: {
+        ffmpegError: err,
+        ffmpegStdout: stdout,
+        ffmpegStderr: stderr,
+      },
+    };
+  }
+  logs.debug("finished transcoding locally");
+
+  logs.debug("uploading transcoded file");
+  const uploadResponse = await bucket.upload(localOutputPath, {
+    destination: storageOutputPath,
+    metadata: { metadata: { isTranscodeOutput: true } },
+  });
+  logs.debug("uploaded transcoded file");
+
+  return {
+    state: "success",
+    sampleRateHertz: streams[0].sample_rate,
+    audioChannelCount: streams[0].channels,
+    uploadResponse,
+    outputPath: storageOutputPath,
+    warnings,
+  };
+}
+
+async function transcodeLocally({
+  inputPath,
```

Comment:
Is there any reason this can't just be `function transcodeLocally(inputPath: string, outputPath: string): Promise<string>`?

Reply:
I don't like having two arguments that are of the same type such that order matters, seems easy to introduce a bug. I prefer having a label for each arg since they're easily accidentally flipped.

Reply:
The method signature becomes unwieldy like in `transcodeToLinear16AndUpload` where some arguments are in an object and others aren't. Every library in the world has methods that have multiple arguments with the same type that can be accidentally flipped; it's the responsibility of the caller to look at the method signature and provide the correct inputs. After all, this doesn't actually stop someone from passing in the wrong data, they would just make a different mistake.

Reply:
TS/JS best practices recommend avoidance of multiple arguments of the same type (DMed you a link since it's internal). To quote, "Strive not to force call site readers to go to the method definition to understand the call". It recommends option bags, which look like what I've got, I think. The ease of making objects in typescript & the availability of destructuring makes the tradeoff more worth it than it would be in, like, Java or C++.

Maybe a better solution would be for me to include `bucket` in the option bag; that might be more consistent. What do you think?

If you do feel strongly about this, I'm willing to go with your suggestion of two string arguments, especially since the function is private.

Reply:
Ah, I see. If it's idiomatic to TS then that's okay. That's what I wanted to confirm. In any case, I think once you have a bag, you should put all the arguments in there, not just some like in the case of `transcodeToLinear16AndUpload`. I think this is consistent with our other OSS work.

Reply:
SGTM, I will do that and get back to you.

Reply:
Done.